comp/forwarder/defaultforwarder/resolver/domain_resolver.go

```diff
@@ -33,38 +38,98 @@ type DomainResolver interface {
 	// Resolve returns the domain to be used to send data for a given `transaction.Endpoint` along with a
 	// destination type
 	Resolve(endpoint transaction.Endpoint) (string, DestinationType)
+	// GetEndpoints returns the list of endpoints associated with this `DomainResolver`
+	GetEndpoints() []utils.APIKeys
 	// GetAPIKeys returns the list of API Keys associated with this `DomainResolver`
 	GetAPIKeys() []string
+	// SetAPIKeys updates the deduped list of API keys associated with this `DomainResolver`
+	SetAPIKeys(keys []string)
 	// GetBaseDomain returns the base domain for this `DomainResolver`
 	GetBaseDomain() string
 	// GetAlternateDomains returns all the domains that can be returned by `Resolve()` minus the base domain
 	GetAlternateDomains() []string
 	// SetBaseDomain sets the base domain to a new value
 	SetBaseDomain(domain string)
 	// UpdateAPIKey replaces instances of the oldKey with the newKey
-	UpdateAPIKey(oldKey, newKey string)
+	UpdateAPIKey(configPath, oldKey, newKey string)
 	// GetBearerAuthToken returns Bearer authtoken, used for internal communication
 	GetBearerAuthToken() string
 }
 
 // SingleDomainResolver will always return the same host
 type SingleDomainResolver struct {
-	domain  string
-	apiKeys []string
-	mu      sync.Mutex
+	domain         string
+	apiKeys        []utils.APIKeys
+	dedupedAPIKeys []string
+	mu             sync.Mutex
+}
+
+// OnUpdateConfig adds a hook into the config which will listen for updates to the API keys
+// of the resolver.
+func OnUpdateConfig(resolver DomainResolver, log log.Component, config config.Component) {
+	config.OnUpdate(func(setting string, oldValue, newValue any) {
+		found := false
+
+		for _, endpoint := range resolver.GetEndpoints() {
+			if endpoint.ConfigSettingPath == setting {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return
+		}
+
+		if strings.Contains(setting, "additional_endpoints") {
+			// Updating additional endpoints don't give us the exact key that has been updated so we reload the whole config section.
+			additionalEndpoints := utils.MakeEndpoints(config.GetStringMapStringSlice(setting), setting)
+			endpoints, ok := additionalEndpoints[resolver.GetBaseDomain()]
+			if !ok {
+				log.Errorf("error: the domain in additional_endpoints changed at runtime for '%s', discarding update.", resolver.GetBaseDomain())
+				return
+			}
+			keys := utils.DedupAPIKeys(endpoints)
+			resolver.SetAPIKeys(keys)
+
+			log.Infof("rotating API key for '%s'", setting)
+
+			return
+		}
+
+		oldAPIKey, ok1 := oldValue.(string)
+		newAPIKey, ok2 := newValue.(string)
+		if ok1 && ok2 {
+			resolver.UpdateAPIKey(setting, oldAPIKey, newAPIKey)
+
+			log.Infof("rotating API key for '%s': %s -> %s",
+				setting,
+				scrubber.HideKeyExceptLastFiveChars(oldAPIKey),
+				scrubber.HideKeyExceptLastFiveChars(newAPIKey),
+			)
+
+			return
+		}
+
+		log.Errorf("new API key for '%s' is invalid (not a string) ignoring new value", setting)
```

Comment:
We could also hit this case if the old value was not a string (like nil for example).
Should we error out if the old API key is invalid ?

Reply:
I've added more logging to capture if old and new has changed. 

I'm not sure if it would be wise to just force set a key if the old is invalid as there can be multiple API keys set and without  knowing the old value we won't know if it is a new api key or a changed one. It would be best to revisit when we know how things will work in future..